File: app/src/main/java/com/elementary/tasks/groups/list/GroupsFragment.java
package com.elementary.tasks.groups.list;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.elementary.tasks.R;
import com.elementary.tasks.core.data.models.Group;
import com.elementary.tasks.core.interfaces.SimpleListener;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.Module;
import com.elementary.tasks.core.view_models.groups.GroupsViewModel;
import com.elementary.tasks.databinding.FragmentGroupsBinding;
import com.elementary.tasks.groups.CreateGroupActivity;
import com.elementary.tasks.navigation.fragments.BaseNavigationFragment;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class GroupsFragment extends BaseNavigationFragment {

    private FragmentGroupsBinding binding;
    private GroupsViewModel viewModel;
    private GroupsRecyclerAdapter mAdapter;

    private SimpleListener mEventListener = new SimpleListener() {
        @Override
        public void onItemClicked(int position, View view) {
            startActivity(new Intent(getContext(), CreateGroupActivity.class).putExtra(Constants.INTENT_ID, mAdapter.getItem(position).getUuId()));
        }

        @Override
        public void onItemLongClicked(int position, View view) {
            String[] items = {getString(R.string.change_color), getString(R.string.edit), getString(R.string.delete)};
            if (mAdapter.getItemCount() == 1) {
                items = new String[]{getString(R.string.change_color), getString(R.string.edit)};
            }
            Dialogues.showLCAM(getContext(), item -> {
                switch (item) {
                    case 0:
                        changeColor(mAdapter.getItem(position));
                        break;
                    case 1:
                        startActivity(new Intent(getContext(), CreateGroupActivity.class)
                                .putExtra(Constants.INTENT_ID, mAdapter.getItem(position).getUuId()));
                        break;
                    case 2:
                        viewModel.deleteGroup(mAdapter.getItem(position));
                        break;
                }
            }, items);
        }
    };

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentGroupsBinding.inflate(inflater, container, false);
        initGroupsList();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViewModel();
    }

    private void initViewModel() {
        viewModel = ViewModelProviders.of(this).get(GroupsViewModel.class);
        viewModel.allGroups.observe(this, groups -> {
            if (groups != null) {
                showGroups(groups);
            }
        });
    }

    private void showGroups(List<Group> groups) {
        mAdapter.setData(groups);
        refreshView();
    }

    private void changeColor(final Group group) {
        String[] items = {getString(R.string.red), getString(R.string.purple),
                getString(R.string.green), getString(R.string.green_light),
                getString(R.string.blue), getString(R.string.blue_light),
                getString(R.string.yellow), getString(R.string.orange),
                getString(R.string.cyan), getString(R.string.pink),
                getString(R.string.teal), getString(R.string.amber)};
        if (Module.isPro()) {
            items = new String[]{getString(R.string.red), getString(R.string.purple),
                    getString(R.string.green), getString(R.string.green_light),
                    getString(R.string.blue), getString(R.string.blue_light),
                    getString(R.string.yellow), getString(R.string.orange),
                    getString(R.string.cyan), getString(R.string.pink),
                    getString(R.string.teal), getString(R.string.amber),
                    getString(R.string.dark_purple), getString(R.string.dark_orange),
                    getString(R.string.lime), getString(R.string.indigo)};
        }
        Dialogues.showLCAM(getContext(), item -> viewModel.changeGroupColor(group, item), items);
    }

    private void initGroupsList() {
        binding.recyclerView.setHasFixedSize(false);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new GroupsRecyclerAdapter(mEventListener);
        binding.recyclerView.setAdapter(mAdapter);
        refreshView();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.groups));
            getCallback().onFragmentSelect(this);
            getCallback().setClick(view -> startActivity(new Intent(getContext(), CreateGroupActivity.class)));
            getCallback().onScrollChanged(binding.recyclerView);
        }
    }

    private void refreshView() {
        if (mAdapter == null || mAdapter.getItemCount() == 0) {
            binding.emptyItem.setVisibility(View.VISIBLE);
            binding.recyclerView.setVisibility(View.GONE);
        } else {
            binding.emptyItem.setVisibility(View.GONE);
            binding.recyclerView.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/groups/list/GroupsFragment.kt
package com.elementary.tasks.groups.list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.elementary.tasks.R
import com.elementary.tasks.core.data.ui.group.UiGroupList
import com.elementary.tasks.core.interfaces.ActionsListener
import com.elementary.tasks.core.utils.ListActions
import com.elementary.tasks.databinding.FragmentGroupsBinding
import com.elementary.tasks.navigation.NavigationAnimations
import com.elementary.tasks.navigation.toolbarfragment.BaseToolbarFragment
import com.github.naz013.analytics.Screen
import com.github.naz013.analytics.ScreenUsedEvent
import com.github.naz013.common.intent.IntentKeys
import com.github.naz013.feature.common.livedata.nonNullObserve
import com.github.naz013.ui.common.Dialogues
import com.github.naz013.ui.common.theme.ThemeProvider
import com.github.naz013.ui.common.view.ViewUtils
import com.github.naz013.ui.common.view.applyBottomInsets
import com.github.naz013.ui.common.view.applyBottomInsetsMargin
import org.koin.androidx.viewmodel.ext.android.viewModel

class GroupsFragment : BaseToolbarFragment<FragmentGroupsBinding>() {

  private val viewModel by viewModel<GroupsViewModel>()
  private var groupsRecyclerAdapter = GroupsRecyclerAdapter()

  override fun inflate(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ) = FragmentGroupsBinding.inflate(inflater, container, false)

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.recyclerView.applyBottomInsets()
    binding.fab.applyBottomInsetsMargin()
    binding.fab.setOnClickListener { addGroup() }
    initGroupsList()
    initViewModel()

    analyticsEventSender.send(ScreenUsedEvent(Screen.GROUPS))
  }

  private fun addGroup() {
    navigate {
      navigate(R.id.editGroupFragment, null, NavigationAnimations.inDepthNavOptions())
    }
  }

  private fun initViewModel() {
    viewModel.allGroups.nonNullObserve(viewLifecycleOwner) { showGroups(it) }
  }

  private fun showGroups(reminderGroups: List<UiGroupList>) {
    groupsRecyclerAdapter.submitList(reminderGroups)
    refreshView()
  }

  private fun changeColor(uiGroupList: UiGroupList) {
    dialogues.showColorDialog(
      activity = requireActivity(),
      current = uiGroupList.colorPosition,
      title = getString(R.string.color),
      colors = ThemeProvider.colorsForSliderThemed(requireContext())
    ) {
      viewModel.changeGroupColor(uiGroupList.id, it)
    }
  }

  private fun initGroupsList() {
    groupsRecyclerAdapter.actionsListener = object : ActionsListener<UiGroupList> {
      override fun onAction(view: View, position: Int, t: UiGroupList?, actions: ListActions) {
        if (t == null) return
        when (actions) {
          ListActions.MORE -> {
            showMore(view, t)
          }

          ListActions.EDIT -> {
            editGroup(t.id)
          }

          else -> {
          }
        }
      }
    }

    if (resources.getBoolean(R.bool.is_tablet)) {
      binding.recyclerView.layoutManager = StaggeredGridLayoutManager(
        resources.getInteger(R.integer.num_of_cols),
        StaggeredGridLayoutManager.VERTICAL
      )
    } else {
      binding.recyclerView.layoutManager = LinearLayoutManager(context)
    }
    binding.recyclerView.adapter = groupsRecyclerAdapter
    ViewUtils.listenScrollableView(binding.recyclerView) {
      if (it) {
        binding.fab.show()
      } else {
        binding.fab.hide()
      }
    }

    refreshView()
  }

  private fun showMore(view: View, t: UiGroupList) {
    var items = arrayOf(
      getString(R.string.change_color),
      getString(R.string.edit),
      getString(R.string.delete)
    )
    if (groupsRecyclerAdapter.itemCount == 1) {
      items = arrayOf(getString(R.string.change_color), getString(R.string.edit))
    }
    Dialogues.showPopup(view, { item ->
      when (item) {
        0 -> changeColor(t)
        1 -> editGroup(t.id)
        2 -> askConfirmation(t)
      }
    }, *items)
  }

  private fun askConfirmation(t: UiGroupList) {
    withContext {
      dialogues.askConfirmation(it, getString(R.string.delete)) { b ->
        if (b) viewModel.deleteGroup(t.id)
      }
    }
  }

  private fun editGroup(id: String) {
    val bundle = Bundle().apply {
      putString(IntentKeys.INTENT_ID, id)
    }
    navigate {
      navigate(
        R.id.editGroupFragment,
        bundle,
        NavigationAnimations.inDepthNavOptions()
      )
    }
  }

  override fun getTitle(): String = getString(R.string.groups)

  private fun refreshView() {
    if (groupsRecyclerAdapter.itemCount == 0) {
      binding.emptyItem.visibility = View.VISIBLE
      binding.recyclerView.visibility = View.GONE
    } else {
      binding.emptyItem.visibility = View.GONE
      binding.recyclerView.visibility = View.VISIBLE
    }
  }
}
